### PR TITLE
Recreation of 'Creative SountFont Technology'

### PR DESCRIPTION
**Pull request by Loobinex**
From the 'Creative' Website:
 
> Dungeon Keeper, developed by Bullfrog Productions, has SoundFont compatible banks built right in. If you don't have an AWE card you're missing out.
> 
> However, if you're lucky enough to own an AWE sound card, you can experience this game at its fullest.

See [this topic](https://keeperklan.com/threads/5168-Can-this-feature-be-revived). Simply put, the original game plays sound effects, to enhance the atmosphere, that only owners of a AWE32 creative sound card can hear. It plays a MIDI-file, using a set of .wav files as instruments, which is plays at different intervals and pitch values to give the impression of background sounds in your dungeon, like creaking doors or horrible screams.

Since nobody has a 1998 sound card anymore, this feature is lost to KeeperFX. I've tried to imitate the feature by including the .wav files into a new [sound.dat](https://mega.nz/#!oMgx1a5a!AFoyAVK8Hxq2YmFCaHZk5IkS9lxTb-GcXjR-mTkxxhw) and writing some code to play these sounds with random intervals and pitch values.

You need the line 'ATMOSPHERIC_SOUNDS = ON' in KeeperFX.cfg to have this feature enabled.

You should hear occasional drops of water falling, and after 3 minutes of gameplay you can start to hear other sounds randomly play every roughly 30 seconds, like screams or rattling chains.

### CODE REVIEW

**Comment by spartahawk** [comment]
Awesome. I'll try this out today.

File: src/bflib_sound.c [hunk 1]
```diff
@@ -755,6 +760,7 @@ void play_atmos_sound(long smpl_idx)
     if (!S3DEmitterIsPlayingSample(Non3DEmitter, smpl_idx, 0))
     {
         S3DAddSampleToEmitterPri(Non3DEmitter, smpl_idx, 0, ATMOS_SOUND_PITCH, ATMOS_SOUND_VOLUME, 0, 3, 8, 0x7FFFFFFE);
+		SYNCDBG(9,"Playing atmos sound %d with pitch %d",(int)smpl_idx,(int)ATMOS_SOUND_PITCH);
```
Comment (StanleySweet) [inline src/bflib_sound.c]: Indent seems off.

File: config/keeperfx.cfg [hunk 2]
```diff
@@ -13,8 +13,10 @@ FRONTEND_RES=640x480x32 640x480x32 640x480x32
 ; Original DK resolutions were FRONTEND_RES=320x200x8 320x200x8 640x480x8
 ; List of in-game resolutions. ALT+R will switch between them.
 INGAME_RES=640x480x32 1024x768x32
-;Original DK resolutions were INGAME_RES=320x200x8 640x400x8
+; Original DK resolutions were INGAME_RES=320x200x8 640x400x8
```
Comment (StanleySweet) [inline config/keeperfx.cfg]: Seems unrelated to that change :)

File: src/config.c [hunk 3]
```diff
@@ -648,6 +657,19 @@ short load_configuration(void)
                 COMMAND_TEXT(cmd_num),config_textname);
           }
           break;
+      case 10: // Atmospheric sound
+          i = recognize_conf_parameter(buf,&pos,len,logicval_type);
+          if (i <= 0)
+          {
+              CONFWRNLOG("Couldn't recognize \"%s\" command parameter in %s file.",
+                COMMAND_TEXT(cmd_num),config_textname);
```
Comment (StanleySweet) [inline src/config.c]: Wrong indent

File: src/sounds.c [hunk 4]
```diff
@@ -316,25 +316,28 @@ void update_player_sounds(void)
...
-					// Roughly every 2 seconds drops sound
-					if ((k % 40) == 0)
+					// No atmos drops the first 30 seconds
+					if (game.play_gameturn > 600)
 					{
-					play_atmos_sound(1013);
+						// Roughly every 2 seconds drops sound
+						if ((k % 40) == 0)
+						{
+						play_atmos_sound(1013);
```
Comment (StanleySweet) [inline src/sounds.c]: Missing indent.

File: src/bflib_sound.c (hunk 1 above)
Comment (Loobinex) [reply src/bflib_sound.c]: It's not though, it's the same as the line above in the source file. Perhaps the github layout pushes it a tab to the right somehow.

File: src/bflib_sound.c [hunk 5]
```diff
@@ -730,6 +732,32 @@ void play_non_3d_sample_no_overlap(long smpl_idx)
     }
 }
 
+void play_atmos_sound(long smpl_idx)
```
Comment (StanleySweet) [inline src/bflib_sound.c]: Indent is messed up below.

File: src/bflib_sound.c [hunk 6]
```diff
@@ -730,6 +732,32 @@ void play_non_3d_sample_no_overlap(long smpl_idx)
...
+{
+	int ATMOS_SOUND_PITCH = (70 + (UNSYNC_RANDOM(9) * 4));
+	int ATMOS_SOUND_VOLUME = 256;
+    if (SoundDisabled)
+        return;
+    if (GetCurrentSoundMasterVolume() <= 0)
+        return;
+    if (Non3DEmitter != 0)
```
Comment (StanleySweet) [inline src/bflib_sound.c]: Could probably use elseifs :)

File: src/config.c (hunk 3 above)
Comment (Loobinex) [reply src/config.c]: This one is copied exactly from the other settings listed above. What do you suggest? Changing all 10 settings?

**Comment by StanleySweet** [comment]
Sorry for the noise, not used to the review process on github :)

File: src/bflib_sound.c (hunk 5 above)
Comment (Loobinex) [reply src/bflib_sound.c]: I'm looking at it, but I don't know what is wrong. Could you please elaborate what you suggest I change?

Comment (Loobinex) [reply src/bflib_sound.c]: But they do? Line 737, 739, 741 and 742 all have the exact same indent. Both on the local source file and here on github.

**Comment by Loobinex** [comment]
I have notepad++ that automatically uses tabs. But since the rest of the source code uses spaces I use spaces as well. Notepad++ shows how many spaces there are.

However, I don't know why, but sometimes what is displayed locally on the source file and on github desktop differs from how it shows up here in the source repository online.

And besides the indentations, how does the code look from a functional perspective, is it effective and efficient or do I do stuff I shouldn't?

File: config/keeperfx.cfg (hunk 2 above)
Comment (Loobinex) [reply config/keeperfx.cfg]: You're right. While updating the file I just added the missing space to make it uniform with the other lines. Is not a functional change, just something snuck in.

File: src/bflib_sound.c (hunk 5 above)
Comment (Loobinex) [reply src/bflib_sound.c]: I see that in the review screen that is not the case, but look here:
https://github.com/Loobinex/keeperfx-unofficial/blob/ManualSoundFont/src/bflib_sound.c#L737

**Comment by spartahawk** [comment]
I'll review the code in a few hours. Sorry for the delay.

Don't waste time on indents guys. It has no impact on the actual function
of the code. Especially if it's a matter of displaying differently in
GitHub.

**Comment by StanleySweet** [comment]
Indeed it's no big deal, however it's harder to read patches which have a
lot of whitespaces changes :)
I haven't looked at the code in ages :) I just popped by because I found it
awesome you guys continued Thomaz' work

Stanislas DOLCINI

Etudiant ESILV - Année 5 (Informatique, Big Data & Objets Connectés) Option
Sécurité
*Tél* : +33 6 84 34 76 71
*Mail* : stanislas.dolcini@gmail.com
<https://www.linkedin.com/in/stanislas-dolcini-36236162>


Le dim. 9 déc. 2018 à 13:51, Doug Hawkes <notifications@github.com> a
écrit :

> I'll review the code in a few hours. Sorry for the delay.
>
> Don't waste time on indents guys. It has no impact on the actual function
> of the code. Especially if it's a matter of displaying differently in
> GitHub.
>

**Comment by Loobinex** [comment]
Sure, always happy to get any help we can get.

**Comment by StanleySweet** [comment]
Great ! Then I shall pop by from time to time to see if I can help you
guys. I get emails each time you guys do a PR so it should be easy.

Stanislas DOLCINI

Etudiant ESILV - Année 5 (Informatique, Big Data & Objets Connectés) Option
Sécurité
*Tél* : +33 6 84 34 76 71
*Mail* : stanislas.dolcini@gmail.com
<https://www.linkedin.com/in/stanislas-dolcini-36236162>

**Comment by spartahawk** [comment]
Everything looks good to me! Very cool.

During gameplay, since the "room noise" cuts in and out as the sounds (especially the drips) start and stop, I felt like it worked much more seamlessly with the game music on. Without it, the dungeon dripping echoes etc cut out a little abruptly. But that's just the nature of the way the sounds were recorded.

It's a great option for players to enjoy. Personally I really like it with the game music playing along with it.

All good for a merge into master.
